Quiz\Crossword: add HTML editor mode to Clues and feedback

vuvanhieu143 opened the pull request:
Hi @timhunt ,
- This change will make clue field into a rich text editor and allow to upload of files in it.
- Add a new feedback field with rich text editor.

Sample UI
![image](https://user-images.githubusercontent.com/42837030/213366522-81ad185f-2ede-4170-ad48-ff32e4fc0427.png)

I also update the code for backup/restore, import/export
- Update automation test (behat and unit).
- Remove unused query in data_preprocessing_words
- Update validation logic due to the editor element being an array.
- Add a select so we can change the editor format in the form.
- Add file access logic for those that can upload files.

timhunt commented:
This change is 95%+ excellent, but then, unfortuantely, are the absolute show-stopper problems which mean that I cannot merge it now.

1. Direct reference to $_POST['clue'] is absolute against the Moodle coding guideslines. It is not necessary to do this, so don't.
2. In fact, despite what you wrote in the comment, the use of definition_after_data should not be needed. It is not needed in qtype_multichoice form, which is very similar.
3. Also, if we were going to do the switchformatgroup thing, then format_text_menu() is the wrong set of choices to have in that menu. That function returns something completely different. If we are going to do this feature, then we shoud just swtich to using the standard Moodle editor element (and, if people don't want to use the editor, they should use the standard option in their user profile to turn it off).

vuvanhieu143 commented:
Hi @timhunt ,
Thank you, I did have some concerns about the switch format button. "Do we really need this since users can change the option in their user profile"
So thankful for you your confirmation on this.
I have removed all the code related to that feature on the second commit.

timhunt commented on renderer.php:
```diff
@@ -78,10 +83,11 @@ public function formulation_and_controls(question_attempt $qa,
                 'value' => $inputvalue,
                 'attributes' => $attributes,
-                'title' => $title,
+                'title' => strip_tags($title),
```
Oh. after studying the code enough, I see this is not a title. This is an aria-label.

Still, why do it like this? The input has a proper label. No need for an aria-label as well. The duplication can only be bad.

timhunt commented on styles.css:
```diff
@@ -166,6 +166,9 @@
 }
+.qtype_crossword-grid-wrapper .wrap-clue .clue-count {
+    margin-right: 3px;
+}
```
Moodle is trying to move away from custom CSS, and instead use Boostrap classes.

However, this issue might not be a good time to try to change this in qtype_crossword. Perhaps we should make a separate issue to fix that?

timhunt reviewed (changes requested):
We are getting this, but there are still things that need to be sorted out.

vuvanhieu143 commented on amd/src/crossword.js:
```diff
@@ -28,42 +28,45 @@ import {CrosswordGrid} from 'qtype_crossword/crossword_grid';
+                '<p dir="rtl" style="text-align: right;"><br></p>',
+                '<p dir="ltr" style="text-align: left;"></p>',
+                '<p dir="rtl" style="text-align: right;"></p>',
+            ];
```
Since we change to allow HTML content for clues, this is a list of cases where we consider a clue is an empty content, so when we click the preview button, it will highlight the row with the color red.

timhunt replied:
But, what HTML is considered empty is an implementation deatail of Atto, which could change at any time (and, as we know, atto is going to go away and be replaced.)

So, this is hugely fragile, and is bound to break in the fututre, but I guess it works for now, so we can live with it.

timhunt replied:
And, haveing this huge const array in the middle of another function is not good code organisation. I woudl at least hide this mess in a htmlIsEmpty function, or something.

vuvanhieu143 commented:
Hi @timhunt ,
+ I have added unit tests before and after the upgrade of new fields and updated the restore logic to always set the default to FORMAT_HTML if they are used old backup.
+ Remove the unused string.
+ Change the title to label, so I need to update behat test.
+ Used bootstrap class instead of custom CSS.
+ Set the display feedback depending on the general feedback setting.

For the second commit, I update the minor coding style and the master branch requires pg13 so I need to update that.

timhunt commented on db/install.xml:
```diff
@@ -32,9 +32,9 @@
         <FIELD NAME="startcolumn" TYPE="int" LENGTH="2" NOTNULL="false" SEQUENCE="false" COMMENT="The vertical index."/>
         <FIELD NAME="startrow" TYPE="int" LENGTH="2" NOTNULL="true" SEQUENCE="false" COMMENT="Horizontal index."/>
-        <FIELD NAME="orientation" TYPE="int" LENGTH="1" NOTNULL="true" DEFAULT="0" SEQUENCE="false" COMMENT="The orientation option."/>
+        <FIELD NAME="orientation" TYPE="int" LENGTH="1" NOTNULL="true" DEFAULT="1" SEQUENCE="false" COMMENT="The orientation option."/>
```
Did you mean to change this default value?

vuvanhieu143 replied:
Sorry about this, it is my mistake, I thought I was changing the clueformat column.

timhunt commented on tests/backup_test.php:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    /**
+     * Test old sq backup data from earlier version.
```
sq here is presumably a copy-paste error.

timhunt commented on tests/backup_test.php:
```diff
@@ -0,0 +1,236 @@
+class backup_test extends \restore_date_testcase {
+
+    /** @var qtype_crossword instance of the question type class to test. */
+    protected $qtype;
```
This is only used in one place. Why bother with setUp and tearDown? Why not just do new qtype_crossword(); in the one place you need it? (Oh, and that tear-down is just unnecessary, by the way.)

timhunt commented on tests/backup_test.php:
```diff
@@ -0,0 +1,236 @@
+        global $DB;
+        $this->resetAfterTest();
+        $this->setAdminUser();
+        if (self::get_moodle_version_major() < $version) {
```
Why is Moodle version relevant here? I'm sure it was in the StudentQuiz test you copied, but this is not StudentQuiz.

vuvanhieu143 replied:
Hi @timhunt , the crossword qtype still supports Moodle 3.11, so our backup of 4.0 is not compatible with the 3.11, so I need to make sure, it is tested on the correct version.

timhunt replied:
OK, but then that needs to be a comment in the code, not just a comment in Github which no-one will ever see again.

timhunt reviewed (changes requested):
Well, I would have liked to merge this, but looking at the code, I cannot.

vuvanhieu143 commented on renderer.php:
```diff
@@ -58,12 +58,12 @@ public function formulation_and_controls(question_attempt $qa,
             }
             $clue = $question->format_text($answer->clue, $answer->clueformat, $qa, 'question', 'clue', $answer->answerid);
-            $title = get_string(
+            $label = format_string(get_string(
```
Hi Tim,
I want to remove unnecessary HTML so the accessibility screen reader tool to read the clue correctly.
Example: When we insert the clue with the editor:
It will include <p></p>
1 Across. <p>where is the Christ the Redeemer statue located in<p>?, 6 letter word" to "BRAZIL
so I use format_string to strip the tags.
Is that clear, I did try to use strip_tags in the previous commit but we have format_string to format the string plain text, so I use it.

timhunt replied:
That is not what format_string is for. Are you trying to do better done by using 'clue' => html_to_text($clue, 0, false), in the array below?

timhunt replied:
Oh! I see you almost got this fixed in the second commit. html_to_text is better than strip_tags. E.g. if there is an impage, it replaces the image with the alt text.

vuvanhieu143 commented:
Hi @timhunt ,
Sorry about all the missing detail that you have pointed out.
I will try to be more careful in the future so it doesn't waste your time.

I have updated and rebased the code with the newest update from the main branch.

There are 2 points, I have not changed yet:
1) I want to make sure the screen reader doesn't read unnecessary HTML tags since the clue is to allow HTML tags.
https://github.com/moodleou/moodle-qtype_crossword/pull/13/commits/26fdc207c5e444988efff65853f07e99174a3748#diff-e82e7ce1e17af09c0cd0f5fee5930434efabb50487fa59a889bc3ba36503c1b8R72
2) I want to keep this function because current our CW type still supports branch M 3.11, so I have added a backup file from 3.11
but the issue is branch 3.11 can't use the 4.0 file so I added this function.
https://github.com/moodleou/moodle-qtype_crossword/blob/26fdc207c5e444988efff65853f07e99174a3748/tests/backup_test.php#L221
so when we run CI for branch 3.11, it will skip all the files that have a version greater than the current branch version.
https://github.com/moodleou/moodle-qtype_crossword/blob/26fdc207c5e444988efff65853f07e99174a3748/tests/backup_test.php#L193

Does this make sense?

timhunt commented on questiontype.php:
```diff
@@ -198,12 +217,27 @@ public function export_to_xml($question, qformat_xml $format, $extra = null): st
+                    $expout .= "        <{$xmlfield} {$format->format($value->{$formatfield})}>\n";
+                    $expout .= '        ' . $format->writetext($value->{$xmlfield});
+                    $expout .= $format->write_files($files);
+                    $expout .= "        </{$xmlfield}>\n";
```
I am looking at this and thinking "Why is there not a helper function $format->write_text_with_files()) with a nice API? like import_text_with_files, but there is not, so this code is right.

timhunt commented on tests/backup_test.php:
```diff
@@ -0,0 +1,226 @@
+        $controller->execute_plan();
+        $controller->destroy();
+
+        return $DB->get_record('question', []);
```
You need to read https://tracker.moodle.org/browse/MDL-62483

This does not work in CodeRunner (or similar) is installed.
